feat(version): Skip repository validation when git is unused

marikaner opened the pull request:
Do not fail version update when git features are not used.

## Description
I am just skipping the check on whether we are @ HEAD when git is not used.

## Motivation and Context
closes #1869 

## How Has This Been Tested?
I tried it with our project, but I did not add any automatic tests. It works in both fixed and locked mode and still throws an error in case git commands are not skipped.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

marikaner commented:
Do you think you will accept this pull request if I fix what you mentioned? I would like to make sure it makes sense to invest time here before I actually do it ;)

marikaner commented:
Thank you for your review, @evocateur. I reverted the changes to the lockfile and will add tests as soon as we agree on the usefulness of this PR ;)

I finally understood your concerns. I misunderstood the purpose of the fixed mode (even though I read your documentation multiple times *) and am now aware that my issue is rather a corner case as you will basically only not need to be on a branch when using `--force-publish` as well as skipping tagging + pushing.
I still think it might be an improvement as it moves lerna's behaviour closer to npm's ([npm does not fail without git](https://docs.npmjs.com/cli/version.html)).

Please take a look at my current implementation and let me know whether this goes into the right direction in your opinion. I just overrode the `requiresGit` method and skipped all branch related actions accordingly.

\* "[Use this if you want to automatically tie all package versions together.](https://github.com/lerna/lerna#fixedlocked-mode-default)" confirmed my intuitive understanding...

marikaner commented:
@evocateur, I added some tests, one is still failing, if will try to get this fixed if I have time this evening, otherwise maybe you can help?

evocateur commented on commands/version/__tests__/version-command.test.js:
```diff
@@ -502,6 +502,89 @@ describe("VersionCommand", () => {
+      expect.assertions(2);
+    });
+
+    it("does not throw for version --force-publish --no-git-tag-version --no-changelog", async () => {
```
In the event no tags are found, collect-updates acts as if you had passed `--force-publish` anyway.

marikaner replied:
Ok, I guess this makes it much simpler. :)

marikaner commented on utils/collect-updates/collect-updates.js:
```diff
@@ -51,8 +49,10 @@ function collectUpdates(filteredPackages, packageGraph, execOpts, commandOptions
   let candidates;
 
   if (!committish || forced.has("*")) {
+    log.info("", `Assuming all packages changed`);
```
👍

marikaner commented:
@evocateur thank you. I removed `forcePublish` and `independent` mode from the conditions for `requiresGit`. Also, I moved the initialization of the command properties to `configureProperties` and adjusted the test conditions accordingly.

marikaner commented:
@evocateur What do you think?

evocateur commented on utils/collect-updates/collect-updates.js:
```diff
@@ -51,8 +49,10 @@ function collectUpdates(filteredPackages, packageGraph, execOpts, commandOptions
   let candidates;
 
   if (!committish || forced.has("*")) {
+    log.info("", `Assuming all packages changed`);
```
I ported this file's changes in 9b8dd9c, please rebase when you can.

marikaner replied:
I just merged the current master and resolved the hereby resulting merge conflict. Did you have anything else in mind? For rebasing I would have to either create a new branch or force push. Anyways, if you squash and merge this shouldn't make a difference.

evocateur commented on commands/version/index.js:
```diff
@@ -43,7 +43,15 @@ class VersionCommand extends Command {
+  }
+
+  configureProperties() {
+    super.configureProperties();
```
I must admit I had a bad reaction when I first saw this `super` call, but now... I really like it! It does a good job of compartmentalizing the operations, as you'd expect a class method to do.

marikaner replied:
I wasn't 100% sure about this either, but it seemed legit :)

evocateur commented:
I'll merge this once you've rebased.

evocateur commented:
Thanks!

marikaner commented:
Wohoo!
